skimage/morphology/footprints.py

```diff
@@ -612,41 +612,166 @@ def _cross(r0, r1, dtype=np.uint8):
+    decomposed : tuple[tuple[ndarray, int], ...]
+        A footprint consisting of multiple array-integer pairs. A tuple whose
+        length is equal to the number of unique structuring elements to apply.
+        Each array has the same dtype as `footprint`.
+
+    Raises
+    ------
+    ValueError
+        If `footprint` isn't symmetric, isn't 2-dimensinonal or isn't of odd
+        length in each dimension.
+    RuntimeError
+        If the difference between the decomposed and original footprint exceeds
+        a given `max_error`. This may be because the given footprint isn't
+        convex or because the decomposition is an approximation.
+        Try checking the given footprint or increase `max_error` if
+        a bigger approximation error is allowed.
+
+    Notes
+    -----
+    This technique was proposed in [1]_ and corresponds roughly to algorithm 1
+    of that publication (some details had to be modified to get reliable
+    operation).
+
+    A convex footprint, in terms of this function, is a footprint that doesn't
+    contain holes, and whose surface doesn't contain notches / inlets.
+
+    References
+    ----------
     .. [1] Li, D. and Ritter, G.X. Decomposition of Separable and Symmetric
            Convex Templates. Proc. SPIE 1350, Image Algebra and Morphological
            Image Processing, (1 November 1990).
            :DOI:`10.1117/12.23608`
+
+    Examples
+    --------
+    >>> import skimage as ski
+    >>> footprint = ski.morphology.footprint_rectangle((3, 9))
+    >>> ski.morphology.footprint_cross_decompose(footprint)
+    ((array([[1, 1, 1, 1, 1, 1, 1, 1, 1]], dtype=uint8), 1),
+     (array([[1],
+             [1],
+             [1]], dtype=uint8),
+      1))
+
+    >>> footprint_ellipse = ski.morphology.ellipse(3, 2)
+    >>> footprint_ellipse
+    array([[0, 1, 1, 1, 1, 1, 0],
+           [1, 1, 1, 1, 1, 1, 1],
+           [1, 1, 1, 1, 1, 1, 1],
+           [1, 1, 1, 1, 1, 1, 1],
+           [0, 1, 1, 1, 1, 1, 0]], dtype=uint8)
+    >>> ski.morphology.footprint_cross_decompose(footprint_ellipse)
+    ((array([[1, 1, 1, 1, 1]], dtype=uint8), 1),
+     (array([[0, 1, 0],
+             [1, 1, 1],
+             [0, 1, 0]], dtype=uint8),
+      1),
+     (array([[1],
+             [1],
+             [1]], dtype=uint8),
+      1))
     """
+    if footprint.ndim != 2:
+        msg = (
+            "Can only decompose footprints with 2 dimensions (for now), "
+            f"got {footprint.ndim}"
+        )
+        raise ValueError(msg)
+    if not np.mod(footprint.shape, 2).all():
+        msg = (
+            f"Can only decompose symmetric footprints with an odd length in "
+            f"each dimension, got `{footprint.shape=}`"
+        )
+        raise ValueError(msg)
+    for dim in range(footprint.ndim):
+        mirrored = np.flip(footprint, axis=dim)
+        if np.any(footprint != mirrored):
+            msg = f"`footprint` isn't symmetric in dimension {dim}"
+            raise ValueError(msg)
```

Comment:
With the addition of the `max_error` argument, this check isn't strictly necessary since a non-convex footprint should usually exceed the allowed error rate. However, we may want to keep this because it provides a clearer warning?